varname.py

```diff
@@ -533,51 +492,79 @@ def _debug(msg: str, frame: Optional[FrameType] = None) -> None:
 
 def _get_frame(
-        caller: int,
-        ignore: Optional[
-            List[Union[ModuleType, Tuple[ModuleType, str]]]
-        ] = None
+        frame: int,
+        ignore: Optional[IgnoreType] = None
 ) -> FrameType:
     """This function makes sure that we get the desired frame,
-    by caller and ignroe.
+    by frame and ignroe.
 
-    The caller is the desired stack we want after excluding the frames we want
+    The frame is the desired stack we want after excluding the frames we want
     to ignore. By default, this module (varname) and the builtin modules will
     be ignored.
 
     The frame_index defaults to 1 to exclude the calls from inside this module.
     """
+    # We need this to check the modules that can't be retrieved
+    # by inspect.getmodule
+    ignore_transformed = [
+        getattr(mod_qname, '__varname_ignore_id__', None)
+        if isinstance(mod_qname, ModuleType)
+        else (getattr(mod_qname[0], '__varname_ignore_id__', None),
+              mod_qname[1])
+        for mod_qname in ignore
+        if not isinstance(mod_qname, FunctionType)
+    ] if ignore else []
+
     # Use loop instead of recursion to avoid creating additional stacks
     try:
         # We are at least skipping 2 frames:
         # one is this function, the other is any varname API that uses this
         # Don't bother to test them, just simply skip.
         frame_index = 2
 
-        while caller > 0:
-            frame = sys._getframe(frame_index)
+        while frame > 0:
+            frameobj = sys._getframe(frame_index)
             # loot at next frame anyway at next iteration
             frame_index += 1
-            module = inspect.getmodule(frame)
-            # exect = executing.Source.executing(frame)
+
+            module = inspect.getmodule(frameobj)
+            try:
+                module_ignore_id = frameobj.f_globals['__varname_ignore_id__']
+            except KeyError:
+                # Use '' instead of None, because we want
+                # module to be matched
+                module_ignore_id = ''
 
             if module is sys.modules[__name__]:
-                _debug('Skipping frame from varname', frame)
+                _debug('Skipping frame from varname', frameobj)
                 continue
 
             if module and module.__name__ in sys.builtin_module_names:
                 # havn't find a way to compose a test for this, skip for now
-                _debug('Skipping builtin module', frame) # pragma: no cover
+                _debug('Skipping builtin module', frameobj) # pragma: no cover
                 continue # pragma: no cover
 
+            if ignore and (module in ignore or
+                           module_ignore_id in ignore_transformed):
+                _debug('Ignored by module', frameobj)
+                continue
+
+            if ignore and (frameobj.f_code in (
+                    getattr(maybe_func, '__code__', '')
```

Comment:
The presence of `__code__` and being a `FunctionType` are not the same thing. I would suggest checking for `__code__` throughout instead of using isinstance to allow passing other things like bound methods.

Reply:
Checking for `__code__` throughout how?

Reply:
As in replace `isinstance(..., FunctionType)` with `hasattr(..., '__code__')` or `isinstance(getattr(ignore_entry, '__code__', None), CodeType)`